include/dali/core/mm/async_pool.h

```diff
@@ -0,0 +1,431 @@
+
+  /**
+   * @brief Try to allocate a block from per-stream free memory
+   *
+   * If the allocation fails, the function returns nullptr.
+   */
+  void *try_allocate(PerStreamFreeBlocks &from, size_t bytes, size_t alignment) {
+    unsigned max_padding = std::min<size_t>(std::max<size_t>(bytes / 16, 16), (1 << 20));
+    const unsigned remainder_alignment = 16;
+    const unsigned min_split_remainder = 16;
+    for (auto it = from.by_size.lower_bound({ bytes, nullptr }); it != from.by_size.end(); ++it) {
+      size_t block_size = it->first;
+      pending_free *f = it->second;
+      char *base = f->addr;
+      char *aligned = detail::align_ptr(base, alignment);
+      size_t front_padding = aligned - base;
+      assert(static_cast<ptrdiff_t>(front_padding) >= 0);
+      // NOTE: block_size - front_padding >= size  can overflow and fail - meh, unsigned size_t
+      if (block_size >= bytes + front_padding) {
+        if (!supports_splitting && block_size - bytes - front_padding > max_padding)
+          return nullptr;  // no point in continuing - the map is sorted
+
+        from.by_size.erase(it);
+        char *block_end = base + block_size;
+        char *remainder = detail::align_ptr(aligned + bytes, remainder_alignment);
+        bool split = supports_splitting && remainder + min_split_remainder < block_end;
+        size_t split_size = block_size;
+        if (split) {
+          // Adjust the pending free `f` so that it contains only what remains after
+          // the block was split.
+          size_t remainder_size = block_end - remainder;
+          f->addr = remainder;
+          f->bytes = remainder_size;
+          f->alignment = remainder_alignment;
+          from.by_size.insert({ remainder_size, f });
+          // The block taken out from the free list is now reduced
+          split_size = remainder - base;
+        } else {
+          remove_pending_free(from, f, false);
+        }
+        if (split_size != bytes) {
+          padded_[aligned] = { split_size,
+                               static_cast<int>(front_padding),
+                               static_cast<int>(alignment) };
+        }
+        return aligned;
+      }
+    }
+    return nullptr;
+  }
+
+  /**
+   * @brief Searches per-stream free blocks to find the most recently freed one.
+   */
+  pending_free *find_first_ready(PerStreamFreeBlocks &free) {
+    SmallVector<pending_free *, 128> pending;
+    int step = 1;
+    pending_free *f = free.free_list.head;
+    while (f) {
+      if (f->ready())
+        break;
+      pending.clear();
+      for (int i = 0; i < step; i++) {
+        f = f->next;
+        if (!f)
+          break;
+        pending.push_back(f);
+      }
+    }
+    if (pending.empty()) {
+      if (f) {
+        assert(f->ready());
+        assert(!f->prev || !f->ready());
+      }
+      return f;
+    }
+    auto it = std::partition_point(pending.begin(), pending.end(), [&](pending_free *f) {
+      return !f->ready();
+    });
```

Comment:
With `int step = 1;` it doesn't do much, right?

Reply:
Fixed.